### PR TITLE
Add webauthn support

### PR DESCRIPTION
**Pull request by MancunianSam**
This adds the three pages needed for webauthn hardware token support.
* The registration page
* The error page - this one is only the template and layout but Keycloak redirects to it so it is needed.
* The sign in page

We need javascript on the pages because keycloak uses the navigator API to authenticate with the tokens. I've written this in Typescript and added in webpack to compile this to javascript for the browser.

### CODE REVIEW

File: themes/tdr/login/src/webauthn/index.ts [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+    for (const authnUseCheck of useCheckElements) {
+
+      allowCredentials.push({
+        id: base64url.toBuffer((authnUseCheck as HTMLInputElement).value),
+        type: 'public-key',
+      })
+    }
+    this.doAuthenticate(allowCredentials)
+  }
+
```
Comment (TomJKing) [inline themes/tdr/login/src/webauthn/index.ts:55]: Unnecessary line break

File: themes/tdr/login/src/webauthn/index.ts [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+              this.setInputValue("userHandle", base64url.encode(new Buffer(response.userHandle)))
+            }
+            const webAuth: HTMLFormElement | null = document.querySelector("#webAuth")
+            if (webAuth) {
+              webAuth.submit()
+            }
+          }
+
+        }
+
```
Comment (TomJKing) [inline themes/tdr/login/src/webauthn/index.ts:110]: Unnecessary line break

File: themes/tdr/login/src/webauthn/index.ts [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+      })
+      .catch((err) => {
+        this.setInputValue("error", err)
+        const webAuth: HTMLFormElement | null = document.querySelector("#webAuth")
+        if (webAuth) {
+          webAuth.submit()
+        }
+      })
+
```
Comment (TomJKing) [inline themes/tdr/login/src/webauthn/index.ts:119]: Unnecessary line break

File: themes/tdr/login/src/webauthn/index.ts [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+            this.setInputValue("authenticatorLabel", labelResult)
+            submitRegisterForm()
+          }
+        }
+      })
+      .catch(function (err) {
+        const errorElement: HTMLInputElement | null = document.querySelector("#error")
+        errorElement!.value = err
+        submitRegisterForm()
+
```
Comment (TomJKing) [inline themes/tdr/login/src/webauthn/index.ts:246]: Unnecessary line break

**Review by TomJKing — COMMENTED** [review]
Looks good. Just some very fussy comments.

Dare I mention the following:
some test for the TS
documentation: the README should be updated to mention compiling the TS. Also we should add some instructions to the TDR User Admin documentation on how to go about setting up the hardware token: https://github.com/nationalarchives/tdr-dev-documentation/blob/master/tdr-admins/tdr-user-administrator.md The admin users are likely to have to walk users through the process, so it would be good if this is documented.

File: themes/tdr/login/src/webauthn/index.ts (hunk 1 above)
Comment (MancunianSam) [reply themes/tdr/login/src/webauthn/index.ts:55]: This has all changed now so I think this is ok.

File: themes/tdr/login/src/webauthn/index.ts [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+            this.setInputValue("credentialId", credentialType.id)
+            if (response.userHandle) {
+              this.setInputValue("userHandle", base64url.encode(new Buffer(response.userHandle)))
+            }
+            const webAuth: HTMLFormElement | null = document.querySelector("#webAuth")
+            if (webAuth) {
+              webAuth.submit()
+            }
+          }
+
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply themes/tdr/login/src/webauthn/index.ts:108]: Same here, this is all different, and for the one below.

**Comment by MancunianSam** [comment]
I've added in some jest tests as well which wasn't difficult at all :smile:

File: README.md [hunk 6]
```diff
@@ -177,7 +177,7 @@ To run, build and test locally:
     * If npm is not installed install [nvm](https://github.com/nvm-sh/nvm#intro) in root directory
```
Comment (TomJKing) [inline README.md:177]: Could you just put a note about running the jest tests in the README. There is a section about running the tests for sbt, so maybe rework that to cover both sorts of tests.

File: jest.config.js [hunk 7]
```diff
@@ -0,0 +1,5 @@
+/** @type {import('ts-jest/dist/types').InitialOptionsTsJest} */
+module.exports = {
+  preset: 'ts-jest',
+  testEnvironment: 'jsdom',
+};
```
Comment (TomJKing) [inline jest.config.js:5]: Missing line break.

File: jest.config.js [hunk 8]
```diff
@@ -0,0 +1,5 @@
+/** @type {import('ts-jest/dist/types').InitialOptionsTsJest} */
```
Comment (TomJKing) [inline jest.config.js:1]: Is this commented out code needed?

File: themes/tdr/login/test/webauthn.test.ts [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+  return key
+}
+
+const options: PublicKeyCredentialCreationOptions = {
+  challenge: Buffer.alloc(0),
+  excludeCredentials: [],
+  pubKeyCredParams: [],
+  rp: {id: "", name: ""},
+  user: {id: Buffer.alloc(0), name: "", displayName: ""}
+
```
Comment (TomJKing) [inline themes/tdr/login/test/webauthn.test.ts:60]: Unnecessary line break.

**Review by TomJKing — COMMENTED** [review]
Looks good. Just some minor comments. Once this is deployed, should probably get some to try this out to make sure it works properly.

File: README.md (hunk 6 above)
Comment (MancunianSam) [reply README.md:177]: I've added in a line about running `npm test` from the root directory.

File: jest.config.js (hunk 8 above)
Comment (MancunianSam) [reply jest.config.js:1]: I guess not. I can't remember why it's there but I've taken it out and it still works so that's good.

**Pull request merged**